tests/TestApp.hs

```diff
@@ -0,0 +1,133 @@
+module TestApp
+  ( NewTest (..)
+  , Test (..)
+  , OK (..)
+  , testApp
+  , testOpenApi
+  ) where
+
+import Prelude
+
+import Control.Lens (at, (&), (.~), (?~))
+import Data.Aeson (FromJSON, ToJSON, encode)
+import Data.HashMap.Strict.InsOrd qualified as IOHM
+import Data.OpenApi (OpenApi, ToSchema, declareSchemaRef)
+import Data.OpenApi qualified as OpenApi
+import Data.OpenApi.Declare (evalDeclare, looks)
+import Data.Proxy (Proxy (..))
+import Data.Text (Text)
+import GHC.Generics (Generic)
+import Network.HTTP.Types.Header (hContentType)
+import Network.HTTP.Types.Method (Method)
+import Network.HTTP.Types.Status
+  ( Status
+  , status200
+  , status201
+  , status404
+  , status405
+  )
+import Network.Wai
+
+newtype NewTest = NewTest
+  { name :: Text
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON, ToJSON, ToSchema)
+
+newtype Test = Test
+  { name :: Text
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON, ToJSON, ToSchema)
+
+newtype OK = OK
+  { ok :: Bool
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON, ToJSON, ToSchema)
+
+testApp :: Application
+testApp request respond =
+  respond $ dispatch (pathInfo request) (requestMethod request)
+
+dispatch :: [Text] -> Method -> Response
+dispatch = \case
+  ["tests"] -> dispatchTests
+  ["tests", x] -> dispatchTest x
+  _ -> const response404
+
+dispatchTests :: Method -> Response
+dispatchTests = \case
+  "POST" -> responseJSON status201 $ OK True
+  _ -> response405
+
+dispatchTest :: Text -> Method -> Response
+dispatchTest name = \case
+  "GET" -> case name of
+    "42" -> responseLBS status200 [] "{}" -- invalid
+    _ -> responseJSON status200 $ Test name
+  _ -> response405
+
+responseJSON :: ToJSON a => Status -> a -> Response
+responseJSON status =
+  responseLBS status [(hContentType, "application/json")] . encode
+
+response404 :: Response
+response404 = responseLBS status404 [] "Not Found"
+
+response405 :: Response
+response405 = responseLBS status405 [] "Bad Method"
+
+testOpenApi :: OpenApi
+testOpenApi = flip evalDeclare mempty $ do
+  newTestRef <- declareSchemaRef $ Proxy @NewTest
+  testRef <- declareSchemaRef $ Proxy @Test
+  okRef <- declareSchemaRef $ Proxy @OK
+
+  looks $ \d ->
+    mempty
+      & OpenApi.components . OpenApi.schemas .~ d
+      & OpenApi.components . OpenApi.requestBodies
+        .~ IOHM.fromList
```

Comment:
Is there no corresponding `declareRequestBodyRef` to use?

Reply:
I'm not totally sure I follow. As far as I understand it, the "declare" functions all just add to a running list of declarations (`DeclareT` is just a writer). Which you can then do what you want with... I add them `components.schemas`. I'm not sure what a `declareRequestBodyRef` would do or how I'd use it here.

Reply:
I'm mainly talking about mitigating using the stringly `OpenApi.Ref (OpenApi.Reference "Test")` below, which I thought was the point of having this `DeclareT` guy.

Reply:
Oh yeah, I noticed that too, but it's not actually unrelated to declarations. It's confusing though...

I picture it something like this:

```yaml
schema:
  Test: # the actual schema for Test

requestBodies:
  Test: Ref "Test"

paths:
  "/tests":
    post:
      requestBody: Ref "Test"
```

The use of `declareSchemaRef` and `DeclareT` only handles schemas:

```yaml
# declarations = do
#   testRef <- declareSchemaRef

schemas: {declarations}

requestBodies:
  Test: {testRef}

paths:
  "/tests":
    post:
      requestBody: Ref "Test"
```

This removes one magic `"Test"` string (the name of the schema itself), but it's unrelated to the choice of the name "Test" for the request-body, which is just a simple magic string (not a schema name).

I could DRY that if we're worried, but it wouldn't be part of any declarations:

```yaml
# declarations = do
#   testRef <- declareSchemaRef
#   let testsBodyRefName = "Test"

schemas: {declarations}

requestBodies:
  {testsBodyRefName}: {testRef}

paths:
  "/tests":
    post:
      requestBody: Ref {testsBodyRefName}
```

But I think that'd be a bit silly.

You might think we could in-line schemas to solve it:

```yaml
# declarations = do
#   testRef <- declareSchemaRef

schemas: []

requestBodies: {declarations}

paths:
  "/tests":
    post:
      requestBody: {testRef}
```

But that doesn't really work for a few reasons: the `requestBodies` element is not the same type `schemas`, and if I use `DeclareT` to declare request bodies, I can't use it to declare schemas for responses. I haven't explored that approach fully though, to be fair.

Reply:
Thanks for that write up! It's weird to have this nice system for declaring reusable schemas but then not be able to use it for reusable request bodies.

I'm not worried about this in the test at all, but do you even need to declare `requestBodies`? It's valid to just inline that schema right into the `requestBody`, right?

Just using a string for the schema might make some sense here too IMO. Although I guess you get to share some types between schema generation and encoding/decoding with this technique though.